src/node_zip.cc

```diff
@@ -0,0 +1,454 @@
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to permit
+// persons to whom the Software is furnished to do so, subject to the
+// following conditions:
+//
+// The above copyright notice and this permission notice shall be included
+// in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN
+// NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+// OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
+// USE OR OTHER DEALINGS IN THE SOFTWARE.
+#include <iostream>
+#include <utility>
+#include <vector>
+
+#include "async_wrap.h"
+#include "env.h"
+#include "env-inl.h"
+#include "node_buffer.h"
+#include "node_external_reference.h"
+#include "node.h"
+#include "util.h"
+#include "zip.h"
+#include "v8.h"
+
+namespace node {
+
+namespace zip {
+
+static char const emptyZipArchive[] = {
+  0x50, 0x4B, 0x05, 0x06,
+  0x00, 0x00, 0x00, 0x00,
+  0x00, 0x00, 0x00, 0x00,
+  0x00, 0x00, 0x00, 0x00,
+  0x00, 0x00, 0x00, 0x00,
+  0x00, 0x00
+};
+
+class ZipArchive final : public AsyncWrap {
+ public:
+  enum InternalFields {
+    kInternalFieldCount = AsyncWrap::kInternalFieldCount,
+  };
+
+  ZipArchive(Environment* env, v8::Local<v8::Object> obj, std::vector<char>&& buf);
+  ~ZipArchive();
+
+  void MemoryInfo(MemoryTracker* tracker) const override;
+
+  static void CheckZip(const v8::FunctionCallbackInfo<v8::Value>& args, ZipArchive*& zip);
+  static void CheckZipEntry(const v8::FunctionCallbackInfo<v8::Value>& args, ZipArchive*& zip, zip_uint32_t& index);
+
+  static void GetEntries(const v8::FunctionCallbackInfo<v8::Value>& args);
+  static void AddEntry(const v8::FunctionCallbackInfo<v8::Value>& args);
+  static void DeleteEntry(const v8::FunctionCallbackInfo<v8::Value>& args);
+  static void ReadEntry(const v8::FunctionCallbackInfo<v8::Value>& args);
+  static void StatEntry(const v8::FunctionCallbackInfo<v8::Value>& args);
+  static void RestatEntry(const v8::FunctionCallbackInfo<v8::Value>& args);
+  static void Digest(const v8::FunctionCallbackInfo<v8::Value>& args);
+
+  int GetFileType(zip_uint64_t index) const;
+
+  std::vector<char> buf_;
+  zip_error_t error_;
+  zip_source_t* source_ = nullptr;
+  zip_t* zip_ = nullptr;
+
+  SET_SELF_SIZE(ZipArchive)
+};
+
+ZipArchive::ZipArchive(Environment* env, v8::Local<v8::Object> obj, std::vector<char>&& buf)
+    : AsyncWrap(env, obj, AsyncWrap::PROVIDER_FILEHANDLE)
```

Comment:
You're not doing any async operations, so why inherit from `AsyncWrap`? (If you do, you'd want to use a different provider constant here)

Reply:
I remember I tried using `BaseObject`, but compilation failed to an abstract `MemoryInfoName` if my memory serves me right - since I didn't know what to put there, and most other places I was looking for reference (`node_zlib.cc` / `node_file.cc`) used `AsyncWrap` I also went with it. What should I use instead?

Reply:
> I remember I tried using `BaseObject`

Which you should :)

> , but compilation failed to an abstract `MemoryInfoName` if my memory serves me right 

Which it also should :)

>  since I didn't know what to put there

You’d probably want to use `SET_MEMORY_INFO_NAME()` and `SET_SELF_SIZE()` in the class definition.